### PR TITLE
[UR][HIP] Enable kernel finalization using comgr

### PR DESCRIPTION
**Pull request by Naghasan**
For kernel fusion support for hip, we need to finalize the kernels using comgr. The patch finalizes tagged binaries during buildProgram before handing it over to the hip runtime.

Initial / related PR https://github.com/intel/llvm/pull/11003

### CODE REVIEW

**Comment by Naghasan** [comment]
@JackAKirk could have a wee look please ? (I think you are one of the code owner) This is the UR part of the patch you approved in the main intel/llvm side

**Comment by JackAKirk** [comment]
> @JackAKirk could have a wee look please ? (I think you are one of the code owner) This is the UR part of the patch you approved in the main intel/llvm side

sure

**Comment by Naghasan** [comment]
Thanks

File: source/adapters/hip/CMakeLists.txt
Comment (JackAKirk) [inline source/adapters/hip/CMakeLists.txt:21]: CC @kbenzie for awareness. This change implies UR_HIP_ROCM_DIR semantic has changed. I guess this might affect the CMAKE gymnastics with regard to pointing unified-runtime/intel-llvm to the correct rocm.

File: source/adapters/hip/program.cpp
Comment (JackAKirk) [inline source/adapters/hip/program.cpp:415]: @jchlanda I was just wondering whether this affects your work on `reqd_work_group_size` metadata?

Comment (jchlanda) [reply source/adapters/hip/program.cpp:415]: Yeap, I've added essentially the same snippet to set the metadata.
I guess we should be OK though, whichever patch gets in first will cause a merge conflict for the other, but that should be easy to resolve.

File: source/adapters/hip/program.cpp
Comment (JackAKirk) [inline source/adapters/hip/program.cpp:114]: This line will never be reached right? because of the above assert being set to false.

Also it would be better to use a backend specific error that then gets automatically passed to a sycl extension, using `setErrorMessage`, like here: https://github.com/intel/llvm/pull/10953/files

File: source/adapters/hip/program.cpp
Comment (jchlanda) [reply source/adapters/hip/program.cpp:415]: Agreed, git will likely resolve it for us. Not to worry.

File: source/adapters/hip/CMakeLists.txt
Comment (kbenzie) [reply source/adapters/hip/CMakeLists.txt:21]: Thanks, this looks more logical to me but I'm not well versed in _finding rocm_ CMake game.

File: source/adapters/hip/program.cpp
Comment (Naghasan) [reply source/adapters/hip/program.cpp:114]: > This line will never be reached right? 

In a normal setup, correct. Hence the `false` here.

In order to be reached, you would need to either enable fusion in intel/llvm and actively disable it in the adapter or use the finalize property in some other patch. Either way, you need to actively change something in the build (cmake changes) or add a patch (which if tested should fail CI) to trigger this.

I can add a `setErrorMessage` if you care about a non sycl focus use of the adapter, otherwise I don't think it is that useful.

File: source/adapters/hip/CMakeLists.txt
Comment (Naghasan) [reply source/adapters/hip/CMakeLists.txt:21]: libs found in `hip/lib` are also found in `lib` via a symlink. But logically, the path is no longer HIP but ROCM (which include ROCM). Hence the simple change here.

File: source/adapters/hip/program.cpp
Comment (JackAKirk) [reply source/adapters/hip/program.cpp:114]: Yeah it's fine to leave it as it is.

**Review by JackAKirk — APPROVED** [review]

**Comment by Naghasan** [comment]
Unless there is more comments, can we have this merged please ?

thanks

**Comment by kbenzie** [comment]
> Unless there is more comments, can we have this merged please ?

Updating UR in intel/llvm is currently blocked by https://github.com/intel/llvm/pull/11312 - I'm hoping to get that resolved today but want to do that before merging any more adapter changed to UR, things are already kinda complicated. I'll add it to the list to merge once things are unblocked.

**Pull request merged**
